### PR TITLE
Deserialize ed25519 keys from hashivault correctly

### LINKED ISSUE #1810: Hashivault ED25519 sign/verify fails
**Description**

Tried using an ed25519 key in a hashicorp vault transit engine with cosign. Failed with "PEM decoding failed" error.

Vault returns an ed25519 public key as the base64 encoding of the raw public key bytes with no further encoding (so no PEM to decode).

**Version**

cosign v2.4.0

### PR DESCRIPTION
**Pull request by stevenjohnstone**
#### Summary

Fixes sign/verify using an ed25519 key stored in Hashicorp Vault's transit engine. Vault returns public keys in different encodings depending on the key type. In the case of ed25519, it base64 encodes the raw public key bytes.

Resolves #1810

### CODE REVIEW

File: pkg/signature/kms/hashivault/client.go [hunk 1]
```diff
@@ -220,17 +221,28 @@ func (h *hashivaultClient) fetchPublicKey(_ context.Context) (crypto.PublicKey,
...
 	}
 
-	strPublicKeyPem, ok := publicKeyPem.(string)
+	strPublicKey, ok := publicKey.(string)
 	if !ok {
 		return nil, fmt.Errorf("could not parse public key pem as string")
 	}
+	// vault returns the key type in the "name" field
+	keyType := keyMap["name"]
+	if keyType == "ed25519" {
```
Comment (bobcallaway) [inline pkg/signature/kms/hashivault/client.go:235]:
```suggestion
	if keyType := keyMap["name"]; keyType == "ed25519" {
```

File: pkg/signature/kms/hashivault/client.go [hunk 2]
```diff
@@ -220,17 +221,28 @@ func (h *hashivaultClient) fetchPublicKey(_ context.Context) (crypto.PublicKey,
...
+		// vault returns ed25519 public keys as base64 encoding of
+		// the raw bytes of the key
+		decodedPublicKey, err := base64.StdEncoding.DecodeString(strPublicKey)
+		if err != nil {
+			return nil, fmt.Errorf("failed to base64 decode ed25519 public key: %s", err)
+		}
```
Comment (bobcallaway) [inline pkg/signature/kms/hashivault/client.go:241]: perhaps worth adding a defensive check to ensure `len(decodedPublicKey) == ed25519.PublicKeySize`?

File: pkg/signature/kms/hashivault/client.go (hunk 1 above)
Comment (stevenjohnstone) [reply pkg/signature/kms/hashivault/client.go]: Changed, thanks :+1:

File: pkg/signature/kms/hashivault/client.go (hunk 2 above)
Comment (stevenjohnstone) [reply pkg/signature/kms/hashivault/client.go:240]: Yes, agree it's worth adding a check here. Thanks!

**Review by bobcallaway — APPROVED** [review]

**Pull request merged**